### PR TITLE
Dag suggestions

### PR DESCRIPTION
**Pull request by 6543**
suggestions for https://github.com/woodpecker-ci/woodpecker/pull/2771

### CODE REVIEW

File: pipeline/frontend/yaml/compiler/dag.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	var curStage *backend_types.Stage
+	var curGroup string
+
+	for _, s := range c.dcs {
+		// create a new stage if current step is in a new group compared to last one
+		if curStage == nil || curGroup != s.group || s.group == "" {
+			curGroup = s.group
+
+			curStage = new(backend_types.Stage)
+			curStage.Name = fmt.Sprintf("%s_stage_%v", c.prefix, s.position)
```
Comment (anbraten) [inline pipeline/frontend/yaml/compiler/dag.go]: Actually pretty interesting that it uses the container position and not the for example `len(stages)` for the index.

File: pipeline/frontend/yaml/compiler/dag.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+type dagCompilerStep struct {
+	step      *backend_types.Step
+	position  int
+	name      string
+	group     string
+	dependsOn []string
+}
+
+type dagCompiler struct {
+	dcs    []*dagCompilerStep
```
Comment (anbraten) [inline pipeline/frontend/yaml/compiler/dag.go]:
```suggestion
	steps    []*dagCompilerStep
```

File: pipeline/frontend/yaml/compiler/dag.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		return dsc.compileByDependsOn()
+	}
+
+	return dsc.compileByGroup()
+}
+
+func (c dagCompiler) compileByGroup() ([]*backend_types.Stage, error) {
+	stages := make([]*backend_types.Stage, 0, len(c.dcs))
+	var curStage *backend_types.Stage
+	var curGroup string
```
Comment (anbraten) [inline pipeline/frontend/yaml/compiler/dag.go]:
```suggestion
	var currentStage *backend_types.Stage
	var currentGroup string
```

**Review by anbraten — APPROVED** [review]

File: pipeline/frontend/yaml/compiler/dag.go (hunk 1 above)
Comment (6543) [reply pipeline/frontend/yaml/compiler/dag.go]: Yes i keept it as is and did not refactor the alg. we can do that later once we released 2.1.0

**Comment by 6543** [comment]
~~@anbraten you have to merge it :)~~

oh no right I can push to the pull branch :)

**Pull request merged**